Dockerize the pipeline

griffin-h opened the pull request:
I wrote a first attempt at a Dockerfile for the pipeline, plus instructions for setup and use. These changes include the `configure` file, which was previously private, but with no database password. They also include an SQL dump that can be used to reproduce the database structure. This `supernova.sql` file will create a `supernova` database user with no password and give it permission to access all tables in the database.

jfrostburke commented:
Just an update that I tested this build this afternoon. After the minor SQL changes I noted in the commit, and rebuilding/installing the pipeline in the docker container (shebang paths were messed up because I had attempted to install on my computer previously) I could successfully reduce data up to the zcat stage.

In the docker container I ingested data using `LCOGTingest.py` from Griffin's [other PR](https://github.com/svalenti/lcogtsnpipe/pull/26), with
`python LCOGTingest.py -u {observe.lco.global username} -p {observe.lco.global password} -l 10 -r reduced -x`
This ingested the 10 most recent images into the db. I tried `-s {psf, checkpsf, psfmag, checkmag}` all successfully, including windows forwarding from the container.

As I said, the zcat stage failed. This is ultimately because `lsc.lscabsphotdef.get_other_filters` returns `{}`. This function is just an sql query that uses the `photlco` and `telescopes` tables, and it is supposed to return a dictionary containing filters observed by the relevant telescope. It failed because it also matches on the `shortname` column of the `telescopes` table, which was `NULL` instead of the usual `COJ 1m` or something like that. When I updated that value to not be `NULL` (or when I deleted the `shortname`-matching line from the query), then the zcat stage ran correctly. I'll fix this in a commit tomorrow after I think about a reasonable solution.

Full error, for posterity:
```
supernova@cfd28c041f80:~/SN2020hvp$ lscloop.py -n SN2020hvp -f sloan -e 20200428 -f sloan -s zcat --field sloan -F
['up', 'SDSS-U', 'gp', 'SDSS-G', 'rp', 'SDSS-R', 'ip', 'SDSS-I', 'zs', 'Pan-Starrs', 'Pan-Starrs-Z', 'w']
1
####################################################################################################
# IMAGE                                    OBJECT           FILTER           WCS             PSF           PSFMAG    APMAG       ZCAT          MAG      ABSCAT
coj1m011-fa12-20200428-0446-e91	   SN2020hvp	       gp	      0.0	coj1m011-fa12-20200428-0446-e91.psf	  -7.9358	   -8.016	        X	   9999.0	        X
coj1m011-fa12-20200428-0447-e91	   SN2020hvp	       rp	      0.0	coj1m011-fa12-20200428-0447-e91.psf	  -7.9309	    -8.01	        X	   9999.0	        X
coj1m011-fa12-20200428-0448-e91	   SN2020hvp	       ip	      0.0	coj1m011-fa12-20200428-0448-e91.psf	  -6.7816	   -6.797	        X	   9999.0	        X

###  total number = 3
WARNING: FITSFixedWarning: The WCS transformation has more axes (2) than the image it is associated with (0) [astropy.wcs.wcs]
____________________________________________________________________________________________________
Calibrating coj1m011-fa12-20200428-0446-e91.fits to SN2020hvp_sloan.cat
UPDATE photlco SET zcat='X' WHERE filename='coj1m011-fa12-20200428-0446-e91.fits'
transform sloan to natural system
un = u - 0.0 * (u-g)
gn = g - 0.109 * (g-r)
rn = r - 0.027 * (r-i)
in = i - 0.036 * (r-i)
zn = z - -0.04 * (i-z)
13.99120046081543 1.0 8.19422852745
UPDATE photlco SET limmag=21.6442916558, zn=23.8690781797, dzn=0.0161878100828, znnum=190 WHERE filename='coj1m011-fa12-20200428-0446-e91.fits'
Traceback (most recent call last):
  File "/usr/local/bin/lscloop.py", line 215, in <module>
    run_absphot(img)
  File "/usr/local/bin/lscloop.py", line 206, in run_absphot
    args.type, args.force, args.show, args.cutmag, args.calib, args.zcatold)
  File "/usr/local/lib/python2.7/site-packages/lsc/lscabsphotdef.py", line 487, in absphot
    colorvec=colors[lsc.sites.filterst1[_filter]]
KeyError: 'g'
```

griffin-h commented:
> As I said, the zcat stage failed. This is ultimately because `lsc.lscabsphotdef.get_other_filters` returns `{}`. This function is just an sql query that uses the `photlco` and `telescopes` tables, and it is supposed to return a dictionary containing filters observed by the relevant telescope. It failed because it also matches on the `shortname` column of the `telescopes` table, which was `NULL` instead of the usual `COJ 1m` or something like that. When I updated that value to not be `NULL` (or when I deleted the `shortname`-matching line from the query), then the zcat stage ran correctly. I'll fix this in a commit tomorrow after I think about a reasonable solution.

Right, I had this problem too but forgot to document it. Don't delete `shortname` from the query because that is used to figure out what color terms to calculate for that night-site. You should just modify `LCOGTingest.db_ingest` to populate that field correctly when creating a new telescope (similar to how Stefano fixed `lsc.mysqldef.ingestredu`). You may need to do the same thing for the `type` column in the `instruments` table

griffin-h commented on Dockerfile:
```diff
@@ -0,0 +1,69 @@
+
+RUN python setup.py install
+
+#RUN ln -s $LCOSNDIR/data/apass/ $LCOSNDIR/data/sloan/ /usr/local/lib/python2.7/site-packages/lsc/standard/cat/
```
can someone explain why this is commented out?

griffin-h replied:
looks like it was from @abostroem's commit

abostroem replied:
When I built the docker image this I got this error message:
```
Removing intermediate container a752516a6ec8
 ---> 8893da0b3c97
Step 21/25 : RUN ln -s $LCOSNDIR/data/apass/ $LCOSNDIR/data/sloan/ /usr/local/lib/python2.7/site-packages/lsc/standard/cat/
 ---> Running in 941c3e3ae4db
ln: failed to create symbolic link '/usr/local/lib/python2.7/site-packages/lsc/standard/cat/apass': File exists
ln: failed to create symbolic link '/usr/local/lib/python2.7/site-packages/lsc/standard/cat/sloan': File exists
The command '/bin/sh -c ln -s $LCOSNDIR/data/apass/ $LCOSNDIR/data/sloan/ /usr/local/lib/python2.7/site-packages/lsc/standard/cat/' returned a non-zero code: 1
```

griffin-h replied:
I think this happens when you have used the git repo previously and there are existing catalogs. My commit below moves these out of the way first and then does the linking.

griffin-h commented:
> As I said, the zcat stage failed. This is ultimately because `lsc.lscabsphotdef.get_other_filters` returns `{}`. This function is just an sql query that uses the `photlco` and `telescopes` tables, and it is supposed to return a dictionary containing filters observed by the relevant telescope. It failed because it also matches on the `shortname` column of the `telescopes` table, which was `NULL` instead of the usual `COJ 1m` or something like that. When I updated that value to not be `NULL` (or when I deleted the `shortname`-matching line from the query), then the zcat stage ran correctly. I'll fix this in a commit tomorrow after I think about a reasonable solution.

For the record, @jfrostburke fixed this in #26.

griffin-h commented:
> rebuilding/installing the pipeline in the docker container (shebang paths were messed up because I had attempted to install on my computer previously)

My latest commit avoids this problem by forcing the package to rebuild before installation.
